# Table of Contents
### 1. Importing Libraries & Data
### 2. Data Consistency Checks - Products
### 3. Data Consistency Checks - Orders
### 4. Exporting Data

# 1. Importing Libraries & Data

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Creating Path
path = r'/Users/mattwawner/Documents/8-2022-Instacart Basket Analysis'

In [3]:
# Importing Orders Data
df_orders = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col = False)

In [4]:
# Importing Products Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

# 2. Data Consistency Checks - Products

In [5]:
df_orders.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [6]:
# Create a dataframe
df_test = pd.DataFrame()

In [7]:
# Create a mixed type column
df_test['mix'] = ['a','b',1,True]

In [8]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [9]:
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type)!= df_test[[col]].iloc[0].apply(type)).any(axis=1) 
    if len (df_test[weird]) > 0:
        print(col)

mix


In [10]:
df_test['mix'] = df_test['mix'].astype('str')

In [11]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [14]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [15]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [16]:
df_prods_clean.shape

(49677, 5)

In [17]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [18]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [19]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [22]:
df_orders = df_orders.drop(columns = ['eval_set'])

In [23]:
df_orders.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0   2539329        1             1          2                  8   
1   2398795        1             2          3                  7   
2    473747        1             3          3                 12   
3   2254736        1             4          4                  7   
4    431534        1             5          4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 3. Data Consistency Checks - Orders

In [24]:
# Run the describe function on df_prods

df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

#### 2. It seems strange that the Count for 'product_id' is 49,693 but the Max is only 49,688. Also, the Max value for 'prices', 99,999, seems unrealistic and should be investigated further.

In [25]:
# 3. Checking for mixed-type data in 'df_orders'

for col in df_orders.columns.tolist():
    weird = (df_orders[[col]].applymap(type) 
            !=df_orders[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_orders[weird]) > 0:
        print(col)

#### 3. No mixed-type data was found in the 'df_orders' dataframe.

In [40]:
# Checking for missing values in 'df_orders'

df_orders.isnull().sum()

order_id                  0
user_id                   0
order_number              0
order_dow                 0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

In [27]:
# Creating a subset of missing values in 'df_orders'

df_orders_nan = df_orders[df_orders['days_since_prior_order'].isnull() == True]

In [28]:
df_orders_nan.head()

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0    2539329        1             1          2                  8   
11   2168274        2             1          2                 11   
26   1374495        3             1          1                 14   
39   3343014        4             1          6                 11   
45   2717275        5             1          3                 12   

    days_since_prior_order  
0                      NaN  
11                     NaN  
26                     NaN  
39                     NaN  
45                     NaN

#### 5. The only missing values I found were in the 'days_since_prior_order' column. Since each value of 'NaN' in this column has a corresponding 'order_number' value of 1, it is safe to assume that these are all first time orders, so no prior order exists. 

In [29]:
# Finding the mean of the 'days_since_prior_order' column

df_orders.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

In [30]:
# Finding the median of the 'days_since_prior_order' column

df_orders.median()

order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
order_dow                       3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [36]:
# Replacing the missing values in 'days_since_prior_order' with median value of 7

df_orders['days_since_prior_order'].fillna(7, inplace = True)

In [38]:
# Creating a new dataframe containing all of the newly replaced values

df_orders_clean = df_orders[df_orders['days_since_prior_order'].isnull() == False]

In [39]:
# Checking if operation was successful

df_orders_clean.shape

(3421083, 6)

#### 6. To address the missing values, I replaced them with the median value of the column, 7 days.

In [41]:
# Checking for duplicates in df_orders

df_orders_dups = df_orders_clean[df_orders_clean.duplicated()]

In [42]:
df_orders_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

#### 7. I was unable to detect any duplicates within df_orders_clean

# 4. Exporting Data

In [43]:
df_prods_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Cleaned_Products_Data.csv'))

In [44]:
df_orders_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'Cleaned_Orders_Data.csv'))